File: sem4/mpp/project-cs/repository/UserDBRepository.cs
using project_cs.domain;

namespace project_cs.repo;

public class UserDbRepository : IUserRepository {
    private static readonly log4net.ILog Log =
        log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

    readonly IDictionary<String, string> _props;

    public UserDbRepository(IDictionary<string, string> props) {
        Log.Info("creating user repo ");
        this._props = props;
    }

    public User Add(User elem) {
        Log.Info("add user");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "INSERT INTO User (username, password) VALUES(@username, @password)";

            var paramUsername = comm.CreateParameter();
            paramUsername.ParameterName = "@username";
            paramUsername.Value = elem.username;
            comm.Parameters.Add(paramUsername);

            var paramPassword = comm.CreateParameter();
            paramPassword.ParameterName = "@password";
            paramPassword.Value = elem.hashedPassword;
            comm.Parameters.Add(paramPassword);
            comm.ExecuteNonQuery();
            
            var comm2 = con.CreateCommand();
            comm2.CommandText = "SELECT last_insert_rowid()";
            var id = (Int64)comm2.ExecuteScalar()!;
            elem.id = (int)id;
            return elem;
        }
    }

    public void Delete(int id) {
        Log.Info("delete User");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "DELETE FROM User WHERE ID=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);

            comm.ExecuteNonQuery();
        }
    }

    public void Update(User elem, int id) {
        Log.Info("update User");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "UPDATE User SET username=@username, password=@password WHERE ID=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = elem.id;
            comm.Parameters.Add(paramId);

            var paramUsername = comm.CreateParameter();
            paramUsername.ParameterName = "@username";
            paramUsername.Value = elem.username;
            comm.Parameters.Add(paramUsername);

            var paramPassword = comm.CreateParameter();
            paramPassword.ParameterName = "@password";
            paramPassword.Value = elem.hashedPassword;
            comm.Parameters.Add(paramPassword);

            comm.ExecuteNonQuery();
        }
    }

    public User? GetById(int id) {
        Log.Info("get User ");
        var con = DbUtils.GetConnection(_props);
        User? u = null;
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT username, password, id FROM User WHERE id=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);
            using (var dataR = comm.ExecuteReader()) {
                if (dataR.Read()) {
                    var username = dataR.GetString(0);
                    var password = dataR.GetString(1);
                    u = new User(username, password);
                    u.id = id;
                }
            }
        }

        return u;
    }

    public ICollection<User> GetAll() {
        Log.Info("get all from User");
        var con = DbUtils.GetConnection(_props);
        IList<User> users = new List<User>();
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT username, password, id FROM User";

            using (var dataR = comm.ExecuteReader()) {
                while (dataR.Read()) {
                    var username = dataR.GetString(0);
                    var password = dataR.GetString(1);
                    var id = dataR.GetInt32(2);
                    var u = new User(username, password);
                    u.id = id;
                    users.Add(u);
                }
            }
        }

        return users;
    }

    public User? GetUserByUsername(string username) {
        Log.Info("get User ");
        var con = DbUtils.GetConnection(_props);
        User? u = null;
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT username, password, id FROM User WHERE username=@username";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@username";
            paramId.Value = username;
            comm.Parameters.Add(paramId);
            using (var dataR = comm.ExecuteReader()) {
                if (dataR.Read()) {
                    var password = dataR.GetString(1);
                    u = new User(username, password);
                    u.id = dataR.GetInt32(2);
                }
            }
        }

        return u;
    }
}

File: sem4/mpp/project-java/repository/src/main/java/repository/UserDBRepository.java
package repository;

import domain.User;
import utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;


public class UserDBRepository implements UserRepository {
	private final JdbcUtils dbUtils;
	private static final Logger logger = LogManager.getLogger(UserDBRepository.class.getName());

	public UserDBRepository(Properties props) {
		logger.info("Initializing user repository with properties: {} ", props);
		dbUtils = new JdbcUtils(props);
	}

	@Override
	public User add(User elem) {
		logger.traceEntry("add user {}", elem);
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("INSERT INTO User (username, password) values (?,?)")) {
			preparedStatement.setString(1, elem.getUsername());
			preparedStatement.setString(2, elem.getHashedPassword());
			int result = preparedStatement.executeUpdate();
			logger.trace("saved {}", result);
			ResultSet resultSet = con.prepareStatement("SELECT last_insert_rowid()").executeQuery();
			resultSet.next();
			elem.setId(resultSet.getInt(1));
			logger.trace("set id {}", elem.getId());
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return elem;
	}

	@Override
	public void delete(Integer id) {
		logger.traceEntry("delete user");
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("DELETE FROM User WHERE id=?")) {
			preparedStatement.setInt(1, id);
			int result = preparedStatement.executeUpdate();
			logger.trace("deleted {}", result);
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
	}

	@Override
	public void update(User elem, Integer id) {
		logger.traceEntry("update user {}", elem);
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("UPDATE User SET username=?, password=? WHERE id=?")) {
			preparedStatement.setString(1, elem.getUsername());
			preparedStatement.setString(2, elem.getHashedPassword());
			preparedStatement.setInt(3, id);
			int result = preparedStatement.executeUpdate();
			logger.trace("saved {}", result);
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
	}

	@Override
	public User getById(Integer id) {
		logger.traceEntry("get user {}");
		Connection con = dbUtils.getConnection();
		User user = null;
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM User WHERE id=?")) {
			preparedStatement.setInt(1, id);
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				if (resultset.next()) {
					String username = resultset.getString("username");
					String password = resultset.getString("password");
					user = new User(username, password);
					user.setId(id);
				} else {
					logger.error("no child with id");
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return user;
	}

	@Override
	public Collection<User> getAll() {
		logger.traceEntry("get users {}");
		Connection con = dbUtils.getConnection();
		List<User> users = new ArrayList<>();
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM User")) {
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				while (resultset.next()) {
					int id = resultset.getInt("id");
					String username = resultset.getString("username");
					String password = resultset.getString("password");
					User user = new User(username, password);
					user.setId(id);
					users.add(user);
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return users;
	}

	@Override
	public User getUserByUsername(String username) {
		logger.traceEntry("get user by username {}");
		Connection con = dbUtils.getConnection();
		User user = null;
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM User WHERE username=?")) {
			preparedStatement.setString(1, username);
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				if (resultset.next()) {
					String password = resultset.getString("password");
					int id = resultset.getInt("id");
					user = new User(username, password);
					user.setId(id);
				} else {
					logger.error("no user with id");
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return user;
	}
}
